### PR TITLE
[Core] porting loop with TLS to use ParallelUtilities

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
This PR ports one non-trivial loop from omp to `ParallelUtilities`

Even though I implemented the ParallelUtilites with TLS I would still very much appreciate if you could take a look ;)
Once we figured out once how to do it properly and we agree we can start also on a bigger scale with the porting

@matekelemen @ashishdarekar this will also be needed for your Hiwi work in the future.

### CODE REVIEW

**Comment by philbucher** [comment]
@roigcarlo there are no direct tests for this are there?

**Comment by loumalouomega** [comment]
TSL or STL?

**Comment by philbucher** [comment]
**T**hread**L**ocal**S**torage

**Comment by loumalouomega** [comment]
> **T**hread**L**ocal**S**torage

image

File: kratos/spatial_containers/bins_dynamic_objects.h [hunk 1]
```diff
@@ -442,20 +442,22 @@ class BinsObjectDynamic {
    * @param MaxNumberOfResults [description]
    */
   void SearchObjectsInRadius(IteratorType const& ThisObjects, SizeType const& NumberOfObjects, std::vector<double>& Radius, std::vector<std::vector<PointerType> >& Results, std::vector<SizeType>& NumberOfResults, SizeType const& MaxNumberOfResults) {
-    PointType Low, High;
-    SearchStructureType Box;
 
-    #pragma omp parallel for private(Low,High,Box)
-    for(int i = 0; i < static_cast<int>(NumberOfObjects); i++) {
-      ResultIteratorType ResultsPointer            = Results[i].begin();
+    using tls_type = std::tuple<PointType, PointType, SearchStructureType>;
```
Comment (RiccardoRossi) [inline kratos/spatial_containers/bins_dynamic_objects.h]: It is the same, but for the sake of understanding i would propose doing it as

~~~cpp
class tls_type 
{
   public:
   PointType Low;
   PointType High;
   SearchStructureType Box;
}

ndexPartition<std::size_t>(NumberOfObjects).for_each(tls_type(), [&](std::size_t i, tls_type& rTLS){
        ResultIteratorType ResultsPointer = Results[i].begin();	        
        ResultIteratorType ResultsPointer = Results[i].begin();
        PointType& Low = rTLS.Low;
        PointType& High = rTLS.High;
        SearchStructureType& Box = rTLS.Box;

~~~

Comment (philbucher) [reply kratos/spatial_containers/bins_dynamic_objects.h]: ok
but I will use a `struct` :)

Comment (RiccardoRossi) [reply kratos/spatial_containers/bins_dynamic_objects.h]: ok, then you can omit the "public" keyword :-)

Comment (philbucher) [reply kratos/spatial_containers/bins_dynamic_objects.h]: done

**Comment by roigcarlo** [comment]
> @roigcarlo there are no direct tests for this are there?

There are some in https://github.com/KratosMultiphysics/Kratos/tree/master/kratos/tests/cpp_tests/spatial_containers

**Review by RiccardoRossi — APPROVED** [review]

**Review by rubenzorrilla — APPROVED** [review]
I'd have used a `tuple` but I think this is matter of personal taste. Seems fine for me too.

**Comment by matekelemen** [comment]
No issues, just a couple questions/suggestions.


Why do we need to pass an instance of TLS in [for_each](https://github.com/KratosMultiphysics/Kratos/blob/095655a7634e492d56db2c3c84e02fbeb049b96f/kratos/utilities/parallel_utilities.h#L193)?
Its only use is to make argument deduction work, and to copy construct local instances for each thread (which means that every type in TLS needs a copy constructor -> unnecessary constraint). At the same time, moving it internally would mean that every TLS type needs a defined move constructor. If ```for_each``` was a member class template instead, we could move TLS into its template parameters and default construct fresh instances of it internally, without calling any copy/move constructors.

If we were to do that, we could also prevent having to define a TLS struct/tuple when calling the loop, and use argument deduction with parameter packs instead. The basic idea wouldn't change, but it would make creating loops much cleaner. Here's an [example](https://godbolt.org/z/73az9P) of what I mean.

(I'm using ```std::apply``` in there for the sake of compactness, which is a C++17 feature. That said, we could easily replace it with syntax that's available for older C++ versions with no extra runtime overhead)

**Comment by RiccardoRossi** [comment]
@matekelemen i tried to read your example, but i think i am a bit lost.

the code you post is not parallel right?

please note than for whatever solution we decide for we have a number of constraints:
1 - we should be able to implement it using openmp or native c++ parallelism (probably not a blocker, simply telling)
2 - we need TLS arguments to be generated once per thread, and not once per item
3 - we should try to avoid arcane errors if something goes wrong in the compilation (i am a bit afraid of the use of "..."even though no doubt it can be useful)

**Comment by RiccardoRossi** [comment]
i'll add that (unfortunately) we are not yet switched to c++17 due to the need of support old centos (with gcc 4.8!!) so for now we need to keep the use of  c++11 until we will be able to upgrade

**Comment by philbucher** [comment]
> No issues, just a couple questions/suggestions.
> 
> Why do we need to pass an instance of TLS in [for_each](https://github.com/KratosMultiphysics/Kratos/blob/095655a7634e492d56db2c3c84e02fbeb049b96f/kratos/utilities/parallel_utilities.h#L193)?
> Its only use is to make argument deduction work, and to copy construct local instances for each thread (which means that every type in TLS needs a copy constructor -> unnecessary constraint). At the same time, moving it internally would mean that every TLS type needs a defined move constructor. If `for_each` was a member class template instead, we could move TLS into its template parameters and default construct fresh instances of it internally, without calling any copy/move constructors.
> 
> If we were to do that, we could also prevent having to define a TLS struct/tuple when calling the loop, and use argument deduction with parameter packs instead. The basic idea wouldn't change, but it would make creating loops much cleaner. Here's an [example](https://godbolt.org/z/73az9P) of what I mean.
> 
> (I'm using `std::apply` in there for the sake of compactness, which is a C++17 feature. That said, we could easily replace it with syntax that's available for older C++ versions with no extra runtime overhead)

@matekelemen cool example! Many things to learn still
I think the main reason that we cannot go fully with the templates and NEED to copy is the use case of `firstprivate`
=> there the TLS is initialized before being passed to the loop and cannot be initialized in the `for_each`
=> there default constructing doesn't work

**Comment by matekelemen** [comment]
Makes sense, I guess there's no avoiding constructor calls then. It's no longer that pretty but we could still get a bit closer to ```omp``` syntax, I'll leave an [example](https://godbolt.org/z/EeMeYe) here for future reference.

**Pull request merged**